### ISSUE TITLE
Function headers for has_ accessors within oneof are put as private

### ISSUE THREAD

**Issue by alfredkcp** [google/protobuf#234]
This bug is found in 3.0.0-alpha-1 and remain unfixed in 3.0.0-alpha-2.

Crosslinking from [Issue 692](https://code.google.com/p/protobuf/issues/detail?id=692&sort=-id&colspec=ID%20Type%20Status%20Priority%20FixedIn%20Owner%20Summary):
> What steps will reproduce the problem?
> 1. define a message with a oneof field with different data types, e.g.
> message test {
>    oneof union_test {
>       string Str   = 1;
>       bool Boolean = 2;
>       int32 Bumber = 3;
>    }
> }
> 2. look for all the has_ accessors inside the generated .pb.h, in this case has_str(), has_boolean() > and has_number().
> 
> What is the expected output? What do you see instead?
> These methods should be listed as public methods inside the test class but somehow there is a  "private:" being inserted instead:
>   ...
>   // optional string Str = 1;
>   private:
>   inline bool has_str() const;
>   public:
>   inline void clear_str();
>   static const int kStrFieldNumber = 3;
>   ...
> 
> What version of the product are you using? On what operating system?
> This happens in 3.0.0-alpha-1 and alpha-2 on Ubuntu 12.04.5 LTS
> 
> Please provide any additional information below.
> - none
>
> This bug should be about how "containing_oneof" wasn't check in src/google/protobuf/compiler/cpp/[cpp_message.cc](https://github.com/google/protobuf/commit/6ef984af4b0c63c1c33127a12dcfc8e6359f0c9e#commitcomment-10125327)

**Comment by pherl** [comment]
For oneofs, get_oneof_case() should be used instead. @cfallin to confirm.

**Comment by cfallin** [comment]
@alfredkcp, I assume you are using a proto with `syntax = "proto3"`, because in proto2, all oneof fields have `has_...()` methods (see `HasHasMethod()`'s first condition, `HasFieldPresence()`, which always returns true for proto1/proto2).

In proto3, this is an intentional design decision: `has_...()` methods go away except for message fields. For oneofs, the idiomatic API pattern is to use `get_oneof_case()` to check which field in the oneof is set.

**Comment by alfredkcp** [comment]
Yes I'm using proto3.

So are we saying moving forward, the only way to check which "oneof" is set in proto3 is through the unionFoo_case() as mentioned by @pherl ?
The current 3.0 alpha behavior is not consistent because there is a has_() for a message whereas the has_() for other datatypes are made private.

**Comment by cfallin** [comment]
That's a good point, thanks. I suppose a message-field-in-a-oneof could be consistent in two different ways: (i) consistent with "all message fields have `has_...()` methods", or (ii) consistent with "no fields in a oneof have `has_...()` methods". It's a matter of subjective design which is better :-)

That said, I just discussed this with @pherl and @xfxyjwf and we've decided to remove the `has_...()` methods for message fields in oneofs, too. We'll make this change in our internal version and it will eventually be down-integrated to Github.
